Nuid

jasper-d opened the pull request:
Notable changes:
- Created a new Nuid implementation (internal for now) because adapting the existing one w/o introducing breaking changes was not desirable in terms of performance and idiomatic (.NET-ish) interface
- Marked the old implementation as obsolete so it can be removed in a future release
- Removed the [different inbox algorithms for old and new requests styles](https://github.com/nats-io/nats.net/blob/master/src/NATS.Client/Conn.cs#L3192) (not sure why the distinction was made) 
- Opted to use base-64 instead of base-62 to avoid modulo bias (cf. https://github.com/nats-io/nuid/issues/15) and gain some performance
- Added a micro benchmark project (uses BenchmarkDotNet) to ensure performance is acceptable (see results below). All performance optimizations are based on evidence provided by benchmark runs.
- Made NATS.Client internals visible to UnitTests and MicroBenchmarks

Let me know if you have questions/change requests

### Edit:
Haven't changed the token generation yet (https://github.com/nats-io/nats.net/issues/351#issuecomment-571816669), would do so in a seperate PR

Issue for reference: #351


### Benchmark Results
NextNuid is in this PR, NUIDNext is the existing (unused) NUID implementation, OldNewInbox is the currently used implementation (https://github.com/jasper-d/nats.net/blob/bd2abf39df3b7c7c06d12ec9a0069e066650f504/src/Benchmarks/MicroBenchmarks/Program.cs).

#### Windows

```
BenchmarkDotNet=v0.12.0, OS=Windows 10.0.18362
Intel Core i7-4712MQ CPU 2.30GHz (Haswell), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.1.101
  [Host]     : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X64 RyuJIT
  Job-JHNFEY : .NET Framework 4.8 (4.8.4075.0), X64 RyuJIT
  Job-LNVMVJ : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X64 RyuJIT


```
|      Method |       Runtime |      Mean |    Error |   StdDev | Ratio |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------ |-------------- |----------:|---------:|---------:|------:|-------:|------:|------:|----------:|
|    NextNuid |    .NET 4.6.2 |  75.57 ns | 0.039 ns | 0.035 ns |  0.17 | 0.0459 |     - |     - |     144 B |
|    NUIDNext |    .NET 4.6.2 | 454.71 ns | 1.150 ns | 1.076 ns |  1.00 | 0.0458 |     - |     - |     144 B |
| OldNewInbox |    .NET 4.6.2 | 789.66 ns | 0.695 ns | 0.616 ns |  1.74 | 0.1040 |     - |     - |     329 B |
|             |               |           |          |          |       |        |       |       |           |
|    NextNuid | .NET Core 3.1 |  81.45 ns | 0.368 ns | 0.344 ns |  0.19 | 0.0458 |     - |     - |     144 B |
|    NUIDNext | .NET Core 3.1 | 436.05 ns | 0.296 ns | 0.277 ns |  1.00 | 0.0458 |     - |     - |     144 B |
| OldNewInbox | .NET Core 3.1 | 586.60 ns | 0.713 ns | 0.667 ns |  1.35 | 0.0706 |     - |     - |     224 B |

#### Linux

```
BenchmarkDotNet=v0.12.0, OS=fedora 31
Intel Core i7-3612QM CPU 2.10GHz (Ivy Bridge), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.1.101
  [Host]     : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X64 RyuJIT
  Job-AEFKXL : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X64 RyuJIT

Runtime=.NET Core 3.1
```
|      Method |     Mean |   Error |  StdDev | Ratio |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------ |---------:|--------:|--------:|------:|-------:|------:|------:|----------:|
|    NextNuid | 116.7 ns | 0.35 ns | 0.33 ns |  0.19 | 0.0458 |     - |     - |     144 B |
|    NUIDNext | 599.7 ns | 0.59 ns | 0.52 ns |  1.00 | 0.0458 |     - |     - |     144 B |
| OldNewInbox | 761.4 ns | 0.62 ns | 0.58 ns |  1.27 | 0.0706 |     - |     - |     224 B |



nuid.go for reference (the figures differ quite a lot between runs, but these are the lowest numbers I got):
``` 
[jd@linux]~/sources/nuid.origin%  go test -bench=.

goos: linux
goarch: amd64
BenchmarkNUIDSpeed-8            12650074                91.9 ns/op            32 B/op          1 allocs/op
BenchmarkGlobalNUIDSpeed-8      11843164               101 ns/op              32 B/op          1 allocs/op
PASS
ok      _/home/jd/sources/nuid.origin   6.192s
```

ColinSullivan1 commented on src/Benchmarks/MicroBenchmarks/NuidBenchmark.cs:
```diff
@@ -0,0 +1,34 @@
+﻿using BenchmarkDotNet.Attributes;
```
A copyright here would be great...

```
// Copyright 2020 The NATS Authors
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
```

jasper-d replied:
Added license headers to `TestNuid.cs` and `Program.cs` as well, hope that's right.

ColinSullivan1 commented on src/NATS.Client/Internals/Nuid.cs:
```diff
@@ -0,0 +1,159 @@
+﻿using System;
```
Copyright would be good...

```
// Copyright 2020 The NATS Authors
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
```

watfordgnf commented on src/NATS.Client/Internals/Nuid.cs:
```diff
@@ -0,0 +1,159 @@
+            {
+                nuidBuffer[i] = (char)_prefix[i];
+            }
+            Monitor.Exit(_nuidLock);
```
This should be in a finally block.

jasper-d replied:
Had this initially (using a `lock` statement actually). Removed it because that has negative performance impact (~13% on .NET FX, negligible on CoreCLR) and I'm pretty sure that the code between Enter and Exit cannot throw (`BitConverter.ToUint64`, `Random.GetNext/NextBytes`, and `RandomNumberGenerator.GetBytes` may in fact throw but not with the inputs used here if the docs are right).

Personally I don't care about the performance impact because I neither use .NET FX nor the old request style (it isn't on a hot path when using the new request style).  So if you feel more comfortable with a `lock` statement I'll change it.

watfordgnf replied:
I'd prefer a `lock` statement or the `Monitor.Exit` within a finally block, my experience has unfortunately not been that exception free code paths stay that way thru framework updates :(

jasper-d replied:
Fair enough, added a lock.

ColinSullivan1 commented:
We dove right in making comments before saying thank you for the PR.  Just wanted to leave a note saying that this is really good and highly appreciated!

jasper-d commented on src/NATS.Client/Internals/Nuid.cs:
```diff
@@ -0,0 +1,159 @@
+            (byte)'g', (byte)'h', (byte)'i', (byte)'j', (byte)'k', (byte)'l', (byte)'m', (byte)'n',
+            (byte)'o', (byte)'p', (byte)'q', (byte)'r', (byte)'s', (byte)'t', (byte)'u', (byte)'v',
+            (byte)'w', (byte)'x', (byte)'y', (byte)'z', (byte)'0', (byte)'1', (byte)'2', (byte)'3', 
+            (byte)'4', (byte)'5', (byte)'6', (byte)'7', (byte)'8', (byte)'9', (byte)'+', (byte)'/',
```
I actually startet with `_` and `-` instead of `+` and `/` because of exactly that special meaning of `/` and because it looks a bit weird in subjects strings. Then switched to RFC 4648 compliant encoding because, well, it's in an IETF RFC at least.

I wonder what's the use case for using a subject string in an URI and wouldn't it be good practice to encode it then anyways?

That beeing said, I don't mind changing it. Thoughts @ColinSullivan1  @watfordgnf ?

kozlovic replied:
But that same RFC has a section for Base64 with alternate alphabet to avoid `/`. It is [here](https://tools.ietf.org/html/rfc4648#section-5), and correspond to your first choice of `-` and `_` (in this order).

jasper-d replied:
Well, good point, never knew that. Done

kozlovic replied:
Thanks!

kozlovic replied:
Seem that you have to fix some tests though, like GetNextNuid_ContainsOnlyValidCharacters

jasper-d replied:
Yes, got ahead of myself. Tests should be green again now.

jasper-d commented:
Thank you, much appreciated. Comments are fair, so it's all good. :)

jasper-d commented:
Sorry, added another commit. Changed most uses of signed integers to unsigned ones. I think this actually improves the code since non-negative integers are expected in multiple places. Also removed a suspicious cast to Int32.

watfordgnf commented:
For posterity, here are the benchmarks on my workstation:

```
BenchmarkDotNet=v0.12.0, OS=Windows 10.0.17763.973 (1809/October2018Update/Redstone5)
Intel Xeon Silver 4114 CPU 2.20GHz, 2 CPU, 20 logical and 20 physical cores
  [Host]     : .NET Framework 4.8 (4.8.4075.0), X64 RyuJIT
  Job-OZBAHI : .NET Framework 4.8 (4.8.4075.0), X64 RyuJIT
  Job-MNUBEA : .NET Core 3.1.0 (CoreCLR 4.700.19.56402, CoreFX 4.700.19.56404), X64 RyuJIT
```

|   Method |       Runtime |      Mean |    Error |   StdDev | Ratio |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------- |-------------- |----------:|---------:|---------:|------:|-------:|------:|------:|----------:|
| NextNuid |    .NET 4.6.2 | 105.72 ns | 0.392 ns | 0.367 ns |  0.23 | 0.0229 |     - |     - |     144 B |
| NUIDNext |    .NET 4.6.2 | 462.81 ns | 1.806 ns | 1.690 ns |  1.00 | 0.0224 |     - |     - |     144 B |
|          |               |           |          |          |       |        |       |       |           |
| NextNuid | .NET Core 3.1 |  92.28 ns | 0.558 ns | 0.466 ns |  0.21 | 0.0200 |     - |     - |     144 B |
| NUIDNext | .NET Core 3.1 | 449.42 ns | 1.683 ns | 1.574 ns |  1.00 | 0.0200 |     - |     - |     144 B |
